gtn/functions/compose.cpp

```diff
@@ -450,3 +456,940 @@ Graph compose(
+  assert(graphDP.ilabels.size() == graphDP.olabels.size());
+  assert(graphDP.ilabels.size() == graphDP.srcNodes.size());
+  assert(graphDP.ilabels.size() == graphDP.dstNodes.size());
+  assert(graphDP.ilabels.size() == graphDP.weights.size());
+
+  const size_t numNodes = graphDP.inArcOffset.size();
+  const size_t numArcs = graphDP.inArcs.size();
+
+  for (size_t i = 0; i < numNodes; ++i) {
+    const int node = graph.addNode(graphDP.start[i], graphDP.accept[i]);
+    assert(node == i);
+  }
+
+  for (size_t i = 0; i < numNodes; ++i) {
+    const int start = graphDP.outArcOffset[i];
+    const int end =
+        (i == (numNodes - 1)) ? numArcs : graphDP.outArcOffset[i + 1];
+
+    for (int j = start; j < end; ++j) {
+      const int dstNode = graphDP.dstNodes[graphDP.outArcs[j]];
+      const int ilabel = graphDP.ilabels[graphDP.outArcs[j]];
+      const int olabel = graphDP.olabels[graphDP.outArcs[j]];
+      const float weight = graphDP.weights[graphDP.outArcs[j]];
+
+      auto newarc = graph.addArc(i, dstNode, ilabel, olabel, weight);
+    }
+  }
+}
+
+Graph compose(const Graph& first, const Graph& second) {
+  GraphDataParallel graphDP1, graphDP2;
+  // Convert from AOS to SOA
+  graphDP1 = convertToDataParallel(first);
+  graphDP2 = convertToDataParallel(second);
+
+  //////////////////////////////////////////////////////////////////////////
+  // Step 1: Data parallel findReachable
+  //////////////////////////////////////////////////////////////////////////
+  std::vector<bool> reachable(first.numNodes() * second.numNodes(), false);
+
+  std::vector<bool> toExplore(first.numNodes() * second.numNodes(), false);
+  std::vector<bool> epsilonMatched(first.numNodes() * second.numNodes(), false);
+
+  const int numNodesFirst = first.numNodes();
+
+  {
+    std::vector<int> acceptDP1 = convertToNodes(graphDP1.accept);
+    std::vector<int> acceptDP2 = convertToNodes(graphDP2.accept);
+
+    for (auto f : acceptDP1) {
+      for (auto s : acceptDP2) {
+        toExplore[TwoDToOneDIndex(f, s, numNodesFirst)] = true;
+        reachable[TwoDToOneDIndex(f, s, numNodesFirst)] = true;
+      }
+    }
+  }
+
+  // This is the outer control loop that would spawn DP kernels
+  while (checkAnyTrue(toExplore)) {
+    // Convert bits set in toExplore to node pairs
+    auto toExploreNodePair = convertToNodePair(toExplore, numNodesFirst);
+
+    // Reset so pristine state for next frontier to explore
+    // No dependence between iterations
+    std::fill(toExplore.begin(), toExplore.end(), false);
+    std::fill(epsilonMatched.begin(), epsilonMatched.end(), false);
+
+    std::vector<int> arcCrossProductOffset;
+    std::vector<std::pair<int, int>> toExploreNumArcs;
+    std::tie(arcCrossProductOffset, toExploreNumArcs) =
+        calculateArcCrossProductOffset(
+            toExploreNodePair, graphDP1, graphDP2, true);
+
+    // If no arcs to process - we are done
+    if (arcCrossProductOffset.empty()) {
+      break;
+    }
```

Comment:
I think check and the two below (L1093 and L1263) will always evaluate to true since `arcCrossProductOffset` is set to have the same size as `toExploreNodePair` which should never be empty when we get here? Maybe make it an assert instead? Alternatively, we could check `totalArcs == 0` but the for loop should handle that efficiently.